storage_table_entity not able to process @odata.type Edm.Int64

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

<!--- Please keep this note for the community --->

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

<!--- Thank you for keeping this note for the community --->


### Terraform Version

1.6.5

### AzureRM Provider Version

3.84.0

### Affected Resource(s)/Data Source(s)

azurerm_storage_table_entity

### Terraform Configuration Files

```hcl
resource "azurerm_storage_table_entity" "test" {
  storage_account_name = "TESTSTG"
  table_name           = "TEST"

  partition_key = "Details"
  row_key       = "11111111111"
  entity = {
    TestId                     = 1
    "TestId@odata.type"        = "Edm.Int64"
  }
}
```


### Debug Output/Panic Output

```shell
Stack trace from the terraform-provider-azurerm_v3.84.0_x5 plugin:
 
panic: interface conversion: interface {} is string, not float64
 
goroutine 67 [running]:
github.com/hashicorp/terraform-provider-azurerm/internal/services/storage.flattenEntity(0xc004128500?)
	github.com/hashicorp/terraform-provider-azurerm/internal/services/storage/storage_table_entity_resource.go:245 +0xc2e
github.com/hashicorp/terraform-provider-azurerm/internal/services/storage.resourceStorageTableEntityRead(0xc004128500, {0x820d2e0?, 0xc00118fb00})
	github.com/hashicorp/terraform-provider-azurerm/internal/services/storage/storage_table_entity_resource.go:182 +0x7e5
github.com/hashicorp/terraform-provider-azurerm/internal/services/storage.resourceStorageTableEntityCreateUpdate(0xc004128500, {0x820d2e0?, 0xc00118fb00})
	github.com/hashicorp/terraform-provider-azurerm/internal/services/storage/storage_table_entity_resource.go:139 +0xc88
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x99ab090?, {0x99ab090?, 0xc00380b050?}, 0xd?, {0x820d2e0?, 0xc00118fb00?})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:766 +0x163
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc001e11a40, {0x99ab090, 0xc00380b050}, 0xc001ad1860, 0xc004128380, {0x820d2e0, 0xc00118fb00})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:909 +0xa7e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc00020eb10, {0x99ab090?, 0xc00380af60?}, 0xc0007eb2c0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/grpc_provider.go:1060 +0xdbc
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc001eb65a0, {0x99ab090?, 0xc00380a570?}, 0xc000c65d50)
	github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/tf5server/server.go:859 +0x56a
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x8d14c00?, 0xc001eb65a0}, {0x99ab090, 0xc00380a570}, 0xc000c65ce0, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:467 +0x169
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0002a81e0, {0x99bca60, 0xc001eb3d40}, 0xc0034339e0, 0xc001f3d110, 0xf54a9f8, 0x0)
	google.golang.org/grpc@v1.58.3/server.go:1374 +0xde7
google.golang.org/grpc.(*Server).handleStream(0xc0002a81e0, {0x99bca60, 0xc001eb3d40}, 0xc0034339e0, 0x0)
	google.golang.org/grpc@v1.58.3/server.go:1751 +0x9e7
google.golang.org/grpc.(*Server).serveStreams.func1.1()
	google.golang.org/grpc@v1.58.3/server.go:986 +0xbb
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 12
	google.golang.org/grpc@v1.58.3/server.go:997 +0x145
 
Error: The terraform-provider-azurerm_v3.84.0_x5 plugin crashed!
 
This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```


### Expected Behaviour

Terraform will finish apply phase successfully

### Actual Behaviour

Terraform apply fails, even the record has been added into the table

### Steps to Reproduce

1. terraform plan will generate (note that TestId value is string now, even configuration uses number
+ resource "azurerm_storage_table_entity" "test" {
      + entity               = {
          + "TestId"            = "1"
          + "TestId@odata.type" = "Edm.Int64"
        }
      + id                   = (known after apply)
      + partition_key        = "Details"
      + row_key              = "11111111111"
      + storage_account_name = "TESTSTG"
      + table_name           = "TEST"
    }
2. terraform apply - fails to finish successfully

### Important Factoids

_No response_

### References

_No response_

--- a/internal/services/storage/storage_table_entity_resource.go
+++ b/internal/services/storage/storage_table_entity_resource.go
@@ -239,10 +239,9 @@ func flattenEntity(entity map[string]interface{}) map[string]interface{} {
 				result[k] = fmt.Sprint(v)
 			case "Edm.Double":
 				result[k] = fmt.Sprintf("%f", v)
-			case "Edm.Int32":
-				fallthrough
-			case "Edm.Int64":
-				result[k] = fmt.Sprintf("%d", int64(v.(float64)))
+			case "Edm.Int32", "Edm.Int64":
+				// `v` returned as string for int 64
+				result[k] = fmt.Sprint(v)
 			case "Edm.String":
 				result[k] = v
 			default:
@@ -264,7 +263,8 @@ func flattenEntity(entity map[string]interface{}) map[string]interface{} {
 					result[k] = fmt.Sprintf("%d", int64(f64))
 					result[k+"@odata.type"] = "Edm.Int32"
 				} else {
-					result[k] = fmt.Sprintf("%f", v)
+					// fmt.Sprintf("%f", v) will return `123.123000` for `123.123`, have to use fmt.Sprint
+					result[k] = fmt.Sprint(v)
 					result[k+"@odata.type"] = "Edm.Double"
 				}
 			case string:

--- a/internal/services/storage/storage_table_entity_resource_test.go
+++ b/internal/services/storage/storage_table_entity_resource_test.go
@@ -280,7 +280,7 @@ resource "azurerm_storage_table_entity" "test" {
     Test             = "Updated"
   }
   lifecycle {
-	ignore_changes = [entity]
+    ignore_changes = [entity]
   }
 }
 `, template, data.RandomInteger, data.RandomInteger)

--- a/internal/services/storage/storage_table_entity_resource_test.go
+++ b/internal/services/storage/storage_table_entity_resource_test.go
@@ -73,7 +73,6 @@ func TestAccTableEntity_update(t *testing.T) {
 func TestAccTableEntity_update_typed(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_storage_table_entity", "test")
 	r := StorageTableEntityResource{}
-
 	data.ResourceTest(t, r, []acceptance.TestStep{
 		{
 			Config: r.basic(data),
@@ -89,6 +88,30 @@ func TestAccTableEntity_update_typed(t *testing.T) {
 			),
 		},
 		data.ImportStep(),
+		{
+			Config: r.updated_typedInt64(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		{
+			Config: r.updated_typedDouble(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		{
+			Config: r.updated_typedString(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		{
+			Config: r.updated_typedBoolean(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
 	})
 }
 
@@ -194,6 +217,89 @@ resource "azurerm_storage_table_entity" "test" {
   entity = {
     Foo              = 123
     "Foo@odata.type" = "Edm.Int32"
+    Test             = "Updated"
+  }
+}
+`, template, data.RandomInteger, data.RandomInteger)
+}
+
+func (r StorageTableEntityResource) updated_typedInt64(data acceptance.TestData) string {
+	template := r.template(data)
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_storage_table_entity" "test" {
+  storage_account_name = azurerm_storage_account.test.name
+  table_name           = azurerm_storage_table.test.name
+
+  partition_key = "test_partition%d"
+  row_key       = "test_row%d"
+  entity = {
+    Foo              = 123
+    "Foo@odata.type" = "Edm.Int64"
+    Test             = "Updated"
+  }
+}
+`, template, data.RandomInteger, data.RandomInteger)
+}
+
+func (r StorageTableEntityResource) updated_typedDouble(data acceptance.TestData) string {
+	template := r.template(data)
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_storage_table_entity" "test" {
+  storage_account_name = azurerm_storage_account.test.name
+  table_name           = azurerm_storage_table.test.name
+
+  partition_key = "test_partition%d"
+  row_key       = "test_row%d"
+  entity = {
+    Foo              = 123.123
+    "Foo@odata.type" = "Edm.Double"
+    Test             = "Updated"
+  }
+}
+`, template, data.RandomInteger, data.RandomInteger)
+}
+
+func (r StorageTableEntityResource) updated_typedString(data acceptance.TestData) string {
+	template := r.template(data)
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_storage_table_entity" "test" {
+  storage_account_name = azurerm_storage_account.test.name
+  table_name           = azurerm_storage_table.test.name
+
+  partition_key = "test_partition%d"
+  row_key       = "test_row%d"
+  entity = {
+    Foo              = "123.123"
+    "Foo@odata.type" = "Edm.String"
+    Test             = "Updated"
+  }
+  lifecycle {
+	ignore_changes = [entity]
+  }
+}
+`, template, data.RandomInteger, data.RandomInteger)
+}
+
+func (r StorageTableEntityResource) updated_typedBoolean(data acceptance.TestData) string {
+	template := r.template(data)
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_storage_table_entity" "test" {
+  storage_account_name = azurerm_storage_account.test.name
+  table_name           = azurerm_storage_table.test.name
+
+  partition_key = "test_partition%d"
+  row_key       = "test_row%d"
+  entity = {
+    Foo              = "true"
+    "Foo@odata.type" = "Edm.Boolean"
     Test             = "Updated"
   }
 }

